feat: save drawings to PDF

Nerixyz opened the pull request:
This adds <kbd>CTRL</kbd>+<kbd>S</kbd> to save drawings and functionality to persist drawings in the `.pdfpc` file. It uses Cairo's PDF backend for rendering.

Both the PDF export and the JSON (`.pdfpc`) export can be configured to happen automatically at exit.
- Set `option save-drawings-on-exit always` to always save the drawings as a PDF (`{pdf-name}-annotated-{date}.pdf`).
- Set `option persist-drawings true` to export the drawing commands to the `.pdfpc` file.

Related issue: https://github.com/pdfpc/pdfpc/issues/437

Nerixyz committed:
feat: save drawings to pdf

fnevgeny commented:
Thanks for your efforts, but I believe the right approach is to save the drawings in the .pdfpc file, not unlike notes, etc., made by the user interactively. Likely formatted as an array representing `drawing_commands`.

nfuhler commented:
Could you elaborate what your intentions with saving the annotations in the .pdfpc file are?
I would like the feature to save the annotations in the pdf file so I can share the annotated pdf afterwards (e.g. share annotated lecture slides with students).

fnevgeny commented:
> Could you elaborate what your intentions with saving the annotations in the .pdfpc file are?

The philosophy of pdfpc is to never alter the presentation file itself. With the annotations kept in the same PDF file, if you wanted to make even a slight modification to a presentation (e.g., to fix a typo), rerunning latex would overwrite the file, and your annotations would be lost.

> I would like the feature to save the annotations in the pdf file so I can share the annotated pdf afterwards (e.g. share annotated lecture slides with students).

Well, share with them also the respective .pdfpc file - and let them know about the existence of pdfpc...

Nerixyz commented:
I can refactor this to save to and load from the `.pdfpc` file. I'd save it as a base64 blob of deflated data in the JSON.

I would still love to be able to have a "renderer" of some sort in pdfpc. Maybe a flag like `-render`, to bake the drawings into the PDF? If not, I'd make a standalone tool to do this, because students often ask about annotated slides.

Sharing the pdfpc file might be the right thing to do from a philosophical perspective. However, the main idea behind sharing annotations is that students can study with them, and pdfpc isn't designed for this (and it shouldn't be). It's designed for presenting slides.

fnevgeny commented:
> I can refactor this to save to and load from the `.pdfpc` file. I'd save it as a base64 blob of deflated data in the JSON.

Why make a large, uneditable binary blob instead of the vector representation?

> I would still love to be able to have a "renderer" of some sort in pdfpc. Maybe a flag like `-render`, to bake the drawings into the PDF? If not, I'd make a standalone tool to do this, because students often ask about annotated slides.

I'm fine with either approach (the latter introduces some code duplication, but on the other hand, it won't have all the GUI libs as dependencies and can be used in various automation scripts).

> Edit: I should also mention that pdfpc wouldn't overwrite the file, it would open a dialog to save it as `{filename}-annotated.pdf`.

So far, pdfpc has no confirmation/warning dialogs at all. I'm not against them in principle, but we should discuss where and when they should be introduced, consistently. Probably it should be done outside of this PR.

fnevgeny commented:
> Then you'd duplicate the object keys. 

First, `x1` and `y1` are equal to `x2` and `y2` of the previous array element, so can be skipped. In fact, we need to store the stroke *path* (with a variable width). Can use the SVG approach to polyline, like:

```xml
<polyline points="60, 110 65, 120 70, 115 75, 130 80, 125 85, 140 90, 135 95, 150 100, 145"/>
```

but using triplets (adding line width).

> One way to "solve" this could be to store them as an array of fixed length, i.e. `[x1, y2, x2, y2, lwidth]` (lwidth can change inside a path). I could try out this second option.

BTW, `xournal` uses a similar approach (in an XML dialect), e.g.:

```xml
<stroke tool="pen" color="black" width="1.41">
163.50 176.25 162.00 174.75 159.75 172.50 156.75 168.75 153.75 165.00 153.00 ...
</stroke>
```
@AndreasBilke, what do you think?

Nerixyz commented:
Implemented the saving/loading to/from `.pdfpc`. While presenting myself, I realized that I wouldn't want to save to `.pdfpc` by default (e.g. doing a test run and forgetting to remove the drawings). I'm not sure which save method should be the default when using <kbd>CTRL</kbd>+<kbd>S</kbd>. For my use-cases, it would always be the PDF to share it with students, while for others it might be saving to the metadata.

halbuer commented:
+1
Would like to have this feature. Why is it still open?

Nerixyz commented:
I rebased the PR on master and made two main changes:

- Added the `persist-drawings` option to configure if drawings are saved to the associated `.pdfpc` file (default: false).
- Removed the `ask` option for `save-drawings-on-exit` because pdfpc never showed a dialog (https://github.com/pdfpc/pdfpc/pull/754#issuecomment-2864485468). Adding this back should be fairly simple.
